### PR TITLE
46 Added getValueConverters(Locale locale) and getByLanguageCode(String languageCode) methods

### LINKED ISSUE #46: Add methods to ValueConverters getByLocale and getByLanguageCode
It is often necessary to return a ValueConverters based on the locale that is set on the system. It would be cool to add the ability to take the necessary converter by locale or language code to ValueConverters.

Like this:
```java
    public ValueConverters getByLocale(Locale locale) {

        if (locale == null) {
            throw new IllegalArgumentException("Locale is null");
        }

        return getByLanguageCode(locale.getLanguage());
    }

    public ValueConverters getByLanguageCode(String languageCode) {

        if (languageCode == null || languageCode.isEmpty()) {
            throw new IllegalArgumentException("Language code is null or empty: " + languageCode);
        }

        switch( languageCode.toLowerCase(Locale.ENGLISH) ){
            case "ru":
                return ValueConverters.RUSSIAN_INTEGER;
            case "cs":
                return ValueConverters.CZECH_INTEGER;
            case "fr":
                return ValueConverters.FRENCH_INTEGER;
            case "de":
                return ValueConverters.GERMAN_INTEGER;
            case "it":
                return ValueConverters.ITALIAN_INTEGER;
            case "uk":
                return ValueConverters.UKRAINIAN_INTEGER;
            case "tr":
                return ValueConverters.TURKISH_INTEGER;
            case "kk":
                return ValueConverters.KAZAKH_INTEGER;
            case "lv":
                return ValueConverters.LATVIAN_INTEGER;
            case "pt":
            case "pt-br":
                return ValueConverters.BRAZILIAN_PORTUGUESE_INTEGER;
            case "pl":
                return ValueConverters.POLISH_INTEGER;
            case "sr":
                return ValueConverters.SERBIAN_INTEGER;
            case "sk":
                return ValueConverters.SLOVAK_INTEGER;
            default:
                return ValueConverters.ENGLISH_INTEGER;
        }
    }
```

### PR DESCRIPTION
**Pull request by altro3**
Fixed #46

### CODE REVIEW

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 1]
```diff
@@ -48,4 +50,52 @@ public String asWords(Integer value) {
 
         return converter.asWords(value);
     }
+
+    public ValueConverters getByLocale(Locale locale) {
+
+        if (locale == null) {
+            throw new IllegalArgumentException("Locale is null");
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:57]: use `verifyNotNull(locale)`, see line 49

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 2]
```diff
@@ -48,4 +50,52 @@ public String asWords(Integer value) {
 
         return converter.asWords(value);
     }
+
+    public ValueConverters getByLocale(Locale locale) {
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:54]: returning `ENGLISH_INTEGER` by default is implicit; it would be better to change this method to `getByLocaleOrDefault(Locale locale, ValueConverters defaultConverter)` and return the passed converter if nothing matches the passed locale

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 3]
```diff
@@ -48,4 +50,52 @@ public String asWords(Integer value) {
...
+
+        if (locale == null) {
+            throw new IllegalArgumentException("Locale is null");
+        }
+
+        return getByLanguageCode(locale.getLanguage());
+    }
+
+    public ValueConverters getByLanguageCode(String languageCode) {
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:63]: change this method to `private`, this way there's only one entry to it and you can remove the null/empty check

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 4]
```diff
@@ -48,4 +50,52 @@ public String asWords(Integer value) {
...
+
+        if (languageCode == null || languageCode.isEmpty()) {
+            throw new IllegalArgumentException("Language code is null or empty: " + languageCode);
+        }
+
+        switch( languageCode.toLowerCase(Locale.ENGLISH) ){
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:69]: This way it's very easy to forget about adding a new switch branch when adding a new language. It'd be better to add a new required field `String languageCode` to `ValueConverters` and then find the converter that has the matching language code.

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 5]
```diff
@@ -21,31 +30,77 @@
...
+        return getByLanguageCode(languageCode, defaultConverters);
+    }
+
+    public static ValueConverters getByLanguageCode(String languageCode, ValueConverters defaultConverters) {
+
+        if (languageCode == null || languageCode.isEmpty()) {
+            throw new IllegalArgumentException("Language code is null or empty: " + languageCode);
+        }
+
+        return BY_LANGUAGE_CODE.getOrDefault(languageCode, defaultConverters);
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:104]: change to:
```
return Arrays.stream(values())
                .filter(it -> it.languageCodes.contains(languageCode))
                .findFirst()
                .orElse(defaultConverter);
```
this way `BY_LANGUAGE_CODE` can be removed, along with the `static` block

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 6]
```diff
@@ -21,31 +30,77 @@
...
+        if (Strings.isNullOrEmpty(locale.getScript())) {
+            languageCode = locale.getLanguage();
+        } else {
+            languageCode = new Locale.Builder().setLanguage(locale.getLanguage()).setScript(locale.getScript()).build().toString();
+        }
+
+        return getByLanguageCode(languageCode, defaultConverters);
+    }
+
+    public static ValueConverters getByLanguageCode(String languageCode, ValueConverters defaultConverters) {
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:98]: rename to `getByLanguageCodeOrDefault`, also rename last arg to `defaultConverter` (singular)

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 7]
```diff
@@ -21,31 +30,77 @@
...
+
+        if (languageCode == null || languageCode.isEmpty()) {
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:100]: change to:
```
verifyNotNull(languageCode);
verify(!languageCode.isEmpty());
```

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 8]
```diff
@@ -21,31 +30,77 @@
...
     }
 
     public String asWords(Integer value) {
         verifyNotNull(value);
 
         return converter.asWords(value);
     }
+
+    public static ValueConverters getByLocale(Locale locale, ValueConverters defaultConverters) {
+
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:85]: remove empty line

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 9]
```diff
@@ -21,31 +30,77 @@
...
+        verifyNotNull(locale);
+
+        String languageCode;
+        if (Strings.isNullOrEmpty(locale.getScript())) {
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:89]: extract method `boolean hasSpecifiedScript(Locale locale)`

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 10]
```diff
@@ -21,31 +30,77 @@
...
+            languageCode = locale.getLanguage();
+        } else {
+            languageCode = new Locale.Builder().setLanguage(locale.getLanguage()).setScript(locale.getScript()).build().toString();
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:92]: extract method `String getLanguageCodeFor(String language, String script)`

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 11]
```diff
@@ -21,31 +30,77 @@
...
+    RUSSIAN_INTEGER(russianContainer().getNumbersConverter(), "ru"),
+    ITALIAN_INTEGER(italianContainer().getNumbersConverter(), "it"),
+    POLISH_INTEGER(polishContainer().getNumbersConverter(), "pl"),
+    CZECH_INTEGER(czechContainer().getNumbersConverter(), "cs"),
+    SLOVAK_INTEGER(slovakContainer().getNumbersConverter(), "sk"),
+    LATVIAN_INTEGER(latvianContainer().getNumbersConverter(), "lv"),
+    KAZAKH_INTEGER(kazakhContainer().getNumbersConverter(), "kk"),
+    UKRAINIAN_INTEGER(ukrainianContainer().getNumbersConverter(), "uk"),
+    SERBIAN_INTEGER(serbianContainer().getNumbersConverter(), Arrays.asList("sr", new Locale.Builder().setLanguage("sr").setScript("Latn").build().toString())),
+    SERBIAN_CYRILLIC_INTEGER(serbianCyrillicContainer().getNumbersConverter(), new Locale.Builder().setLanguage("sr").setScript("Cyrl").build().toString()),
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:45]: thank you for the in-depth research on this!

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java (hunk 3 above)
Comment (MaciejLipinski) [reply src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java]: alright, I agree it might be useful too.

File: src/test/groovy/pl/allegro/finance/tradukisto/ValueConvertersTest.groovy [hunk 12]
```diff
@@ -52,21 +57,122 @@ class ValueConvertersTest extends Specification {
...
+    def "getByLocale for French"() {
+        expect:
+        getByLocale(Locale.FRENCH, null).asWords(1_234) == "mille deux cent trente-quatre"
+    }
+
+    def "getByLocale for Turkish"() {
+        expect:
+        getByLocale(new Locale("tr"), null).asWords(1_234) == "Bin İki Yüz Otuz Dört"
+    }
+
```
Comment (MaciejLipinski) [inline src/test/groovy/pl/allegro/finance/tradukisto/ValueConvertersTest.groovy:177]: Add test that shows `VerifyException` is thrown when `null` or `""` is supplied

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java [hunk 13]
```diff
@@ -13,39 +20,76 @@
...
+        if (hasSpecifiedScript(locale)) {
+            languageCode = locale.getLanguage();
+        } else {
+            languageCode = getLanguageCodeFor(locale.getLanguage(), locale.getScript());
+        }
+
+        return getByLanguageCodeOrDefault(languageCode, defaultConverter);
+    }
+
+    private static boolean hasSpecifiedScript(Locale locale) {
```
Comment (MaciejLipinski) [inline src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:78]: the name suggests inverse behavior. Negate the expression and change order in the if/else, or rename this method to `doesNotHaveSpecifiedScript` (but I think negating and changing if will be nicer)

**Review by MaciejLipinski — CHANGES_REQUESTED** [review]
almost there, one last thing I found

File: src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java (hunk 13 above)
Comment (altro3) [reply src/main/java/pl/allegro/finance/tradukisto/ValueConverters.java:78]: Fixed

**Review by MaciejLipinski — APPROVED** [review]
